DFE-Digital/schools-experience
# Subject specific dates SE-1804

### Context
Schools want to be able to limit what subjects are available on a particular date.

### Changes proposed in this pull request
* Placement dates now have a flag `supports_subjects?` Placement dates for primary schools do not support subjects, placement dates for secondary and colleges do.
Schools that are both primary and secondary are now given an option to identify a placement date as supporting subjects.

* Candidate registration wizard now includes conditional branching depending on whether a step is required for the current school they are applying for. Eg for a school with fixed dates the wizard now starts on the select placement date screen. For flexible date schools they start on the personal information page.

* Candidates applying to a school with fixed dates will now be presented with a list of dates. If subjects are associated with dates candidates will be able to choose the subject/date combination. Primary dates are separately listed on the same page.

* bookings_placement_date_id has moved nesting in the registration_session, it is now under subject_and_date_information. We can use this key to identify in flight registration_sessions and map them to a non subject specific date see `RegistrationSession#subject_and_date_information`

* fixed dates have been moved from the right hand column of the school profile page to the main column and are placed above the 'Start request' button

* the chosen subject has been added to the check your answers screen

### Guidance to review
Manual testing steps:
From the school dashboard, click manage dates and set up a school with subject specific dates. Apply to this school, accept the placement request as the school expect the subject selected in the confirm booking screen to match the subject for the date you applied to.
Switch the school to flexible dates, make sure applying to the school still functions as before.
Repeat this test for primary, secondary, and primary + secondary.
